### ISSUE TITLE
Build error in Apache Royale JS project when project name contains "-"

### ISSUE THREAD

**Issue by Nisaba** [prominic/Moonshine-IDE#288]
Please not allow "-" char in project name for Apache royale project : 
Here is the build error : 

GRAVE: C://Tag-JS/Inscription-Tag/bin/js-debug/Inscription-Tag.js:97: ERROR - Parse error. invalid assignment target
: Inscription-Tag = function() {
:                         ^
: 
: mai 29, 2018 7:06:46 PM com.google.javascript.jscomp.LoggerErrorManager printSummary
: AVERTISSEMENT: 1 error(s), 0 warning(s)

**Comment by piotrzarzycki21** [comment]
Maybe this is because I have made it less restricted in that issue #188. I will definitely think more about that in the next versions.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine can you shed some light on that [line](https://github.com/prominic/Moonshine-IDE/blob/622bc8083c005960c2731e9da4fa77407f91a506/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/as3project/CreateProject.as#L379). If I understand correctly you are preventing from typing all of that characters in the name ? For example I should not able to write "-" or "@" - yes ?

**Comment by rat-moonshine** [comment]
I need to check, @piotrzarzycki21 . It was suppose not to enter any special character (in fact some special characters break file system in different OS) but I see I can do that now. Thank you for pointing my attention to this. I need to revisit once the thread where we discussed this in past.

I shall let you know.

**Comment by piotrzarzycki21** [comment]
I believe it's working only for some of the character from the list. Those characters looks like to be valid not allow them to the project name. This issue is definitely a Bug - we should not allow not only type all of those characters, but also do not allow type hyphen/dash ("-"). 

I think we should just more work on that list to make restrict working. You confirm that all of those character should not be placed I can spent some time on that issue.

**Comment by rat-moonshine** [comment]
We were discussing about this last night, and I was mainly wondering how this remain un-noticed from our tests after #188 changes to input restrictions. 

I still believe allowing specific characters would be more safe to input against project naming. The current regexp restriction as applied (^\\\\\\/?:"|<>*!@#$%^&*()+{}[]:;~, note - this is the more error free version that @JoelProminic  found - ^\\\/?:"|<>*!@#$%\\^&*()+{}[]:;~\'`=) but this still error-prone. There are numerous other special characters in ASCII character set which can be entered if not recorded to the added list - this is overkill. 

I understand you @piotrzarzycki21 wanted to give user ability to input more special characters which might be legit when in other language - which is good thought. But in current scenario, having the specific characters as regexp to restrict seems break-prone. In that scenario, allowing specific characters only as it was before (a-zA-Z0-1._) should serve the purpose without breaking or doing more to manage this.

**Comment by piotrzarzycki21** [comment]
Yes, but in that case you are forcing user to use English oriented languages. Other IDEs allows you to use for example #188. I'm not sure what to do in that case. 

From one point of view we could have even more problems leaving that regexp, but on the other hand who know whether some day Moonshine won't be IDE which supports also JAVA or other languages ? Where maybe naming project using latin type of languages is quite common ?

**Comment by rat-moonshine** [comment]
We will be happy if we found a way to support #188 in future and we can't port a version to release when we know something can break.

**Comment by piotrzarzycki21** [comment]
This code is already in the release 1.12.0.

**Comment by rat-moonshine** [comment]
Yes, this is unfortunate that we didn't gave much attention before a release.

@JoelProminic suggested, that if we still want to use the 'restrict' model then the regexp needs to be fixed at least. He suggested following regexp against the current one:

`^\\\/?:"|<>*!@#$%\\^&*()+{}[]:;~\'=`

Note, this does not include `-` sign, perhaps another `\\-` might do the job. Please, check.

**Comment by piotrzarzycki21** [comment]
I need to work on that more. It looks like restrict doesn't take Regex. I'm moving that to the next release.

**Comment by rat-moonshine** [comment]
If you plan not to work on this in this release, then please return the old 'allowed' mechanism at least that won't break the name input policy (as we had in 1.12.0).

**Comment by piotrzarzycki21** [comment]
I just didn't change anything in the mechanism and I don't won't to. Let's leave it as is. Actually to what mechanism do you refer ? I will work on that for 1.14.0.

**Comment by rat-moonshine** [comment]
The current UI accepts all these special characters which can breaks a project from compilation which is critical. It doesn't matter that we released with 1.12.0 and also doesn't mean we won't fix this now when we have a chance.

~^&*()_+=-`][{}';:.

Above are the special characters which are currently accepts against a project naming. 

If the restriction needs more time to build carefully, then I'd suggest not to allow to entering them by an user - by allowing only specific characters like a-zA-Z0-1.-. This will not fulfill anything towards #188 but the current scenario is not helping to the issue either. It just inject a bug which can breaks the application.

**Comment by piotrzarzycki21** [comment]
> ~^&*()_+=-`][{}';:.
> 
> Above are the special characters which are currently accepts against a project naming.
> 

Just to be clear we shouldn't allow for all of that chars ?

**Comment by piotrzarzycki21** [comment]
I have spent on that some time today an was able to restrict following symbol. @atingupta2005 could you test ? 
`]<[>"{}?!@$+=()*#&~%;:\\\/-`

**Comment by rat-moonshine** [comment]
@JoelProminic found a few keywords were not restricted by updated restriction rules:

`^_\|',./

Its now updated.

**Comment by JoelProminic** [comment]
Some additional details:

Here is the documentation for TextInput.restrict.  The format is not a regular expression, but it does have similarities:  https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/mx/controls/TextInput.html#restrict


I created a new restriction string including the special character formats and had @rat-moonshine test and commit the change.  I used the order on my keyboard to make it easier to keep track of which characters were included in the string.
```
    ^ ~`!@#$%\\^&*()\\-+=[{]}\\\\|:;\'",<.>/?
```

With this restrict string, '_' is the only allowed special characters.  Spaces and '-' are restrictedd as well.
